chore: python project migration to uv

neo-jesse opened the pull request:
# Migrate Python dependency management from Pipenv to UV

## Summary
Migrates `api` and `shared-data` projects from `pipenv` to `uv` for faster dependency resolution and installation. All other projects continue using `pipenv`.

## Key Changes for Api and shared-data
- Migrated `api` and `shared-data` from `Pipfile` to `pyproject.toml` with `uv.lock`
- Created `scripts/python-uv.mk` for UV projects; `scripts/python.mk` remains for pipenv projects
- Updated GitHub Actions setup action to detect project type (UV vs pipenv) and install dependencies accordingly
- Fixed dependencies: added `opentrons-hardware[FLEX]` to api dev deps, `numpy` to shared-data main deps
- Fixed mypy errors (redundant casts, unused ignores, untyped decorators)
- Migrated JSON schema validation from `jsonschema.RefResolver` to `referencing` library
- Refactored GitHub Actions Python setup into modular actions (`install-uv`, `cache-uv`, `setup`)
- Fixed Windows PATH issues for `uv` in Make commands
- Added `install-dev-deps: 'true'` to workflows that run lint/test for UV projects

## Changes for Pipenv Projects
- Updated all pipenv Makefiles to use `pipenv sync --dev` to ensure dev dependencies are installed
- Fixed workflows to properly install dev dependencies for lint/test jobs

## Benefits
- Faster dependency resolution and installation for `api` and `shared-data`
- Standardized `pyproject.toml` configuration for migrated projects
- Improved CI reliability with proper caching and dev dependency handling
- Gradual migration path - other projects can be migrated incrementally

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,118 @@ inputs:
+      id: project-type
+      shell: bash
+      run: |
+        # Only api and shared-data use UV for now
```
can we probe for the existence of a `uv.lock` file instead, or even cut over the makefile

neo-jesse replied:
More than just the 2 projects we are looking at have uv.lock, it's not a good indicator.

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,118 @@ inputs:
+        PROJECT_TYPE="${{ steps.project-type.outputs.type }}"
+        if [ "$PROJECT_TYPE" == "uv" ]; then
+          echo "::notice::Using UV for dependency management"
+          # Add linux extra if on Linux and project has linux optional dependency (for systemd-python in performance-metrics)
```
i don't think this is how we should handle linux-only requirements. python dependency specifiers already have the `sys_platform` qualifier to make something install only on certain operating systems; we should use that depspec rather than having to handle things here.

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,118 @@ inputs:
+            DEV_EXTRA="--extra dev"
+          fi
+          if [ -f "${{ inputs.project }}/uv.lock" ]; then
+            # Use --frozen only if no extras are requested (extras may not be in lockfile)
```
wait why would they not be

neo-jesse replied:
Removed the conditional that skipped --frozen when extras were requested
Now always uses: uv sync --frozen $DEV_EXTRA when lockfile exists
If extras aren't in the lockfile, uv will fail and we'll know

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -120,7 +122,11 @@ jobs:
       - name: Ensure assets build
-        run: make -C api sdist wheel
+        run: |
+          # Ensure uv is in PATH (should already be there from setup action via $GITHUB_PATH)
```
let's rely on the action to do this rather than add redundant code here that has to be maintained

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -94,16 +95,24 @@ jobs:
       - name: Test
-        run: make -C shared-data test-py
+        run: |
+          # Ensure uv is in PATH (should already be there from setup action via $GITHUB_PATH)
```
let's make sure this happens in the action rather than also do it here

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -94,16 +95,24 @@ jobs:
       - name: Ensure assets build
-        run: make -C shared-data dist-py
+        run: |
+          # Ensure uv is in PATH (should already be there from setup action via $GITHUB_PATH)
```
let's make sure this happens in the action rather than also do it here

sfoster1 commented on api/Makefile:
```diff
@@ -66,14 +66,36 @@ all: clean wheel
-	$(pipenv) sync $(pipenv_opts)
-	$(pipenv) run pip freeze
+	@UNAME_S=$$(uname -s 2>/dev/null || echo ""); \
+	LINUX_EXTRA=""; \
```
let's rely on python depspecs rather than do this here. also, multiline make commands are way easier to parse if they're done in a `define...endef` block outside a recipe body

sfoster1 commented on api/Makefile:
```diff
@@ -66,14 +66,36 @@ all: clean wheel
-	$(pipenv) run pip uninstall -y python-can
-	$(pipenv) run pip freeze
+	@UNAME_S=$$(uname -s 2>/dev/null || echo ""); \
+	LINUX_EXTRA=""; \
```
rely on depspecs

sfoster1 commented on api/mypy.ini:
```diff
@@ -3,10 +3,29 @@ plugins = pydantic.mypy, decoy.mypy, numpy.typing.mypy_plugin
 warn_required_dynamic_aliases = True
 warn_untyped_fields = True
 
+# Ignore optional/local packages that may not be installed
```
this is a really big code smell - we shouldn't have to do this, and we really want to have the projects set up during typechecking so that these work. how can we get rid of these?

sfoster1 commented on api/mypy.ini:
```diff
@@ -3,10 +3,29 @@ plugins = pydantic.mypy, decoy.mypy, numpy.typing.mypy_plugin
+[mypy-performance_metrics.*]
+ignore_missing_imports = True
+
+# Test-only dependencies that may not have type stubs
```
why do we need these now?

neo-jesse replied:
If packages aren't installed, mypy would report missing imports
The ignores prevent those errors

sfoster1 commented on shared-data/mypy.ini:
```diff
@@ -20,3 +20,10 @@ warn_return_any = False
 [mypy-opentrons_shared_data.protocol.*]
 warn_return_any = False
+
+# PIL/Pillow is only used in tools, not in main code
```
why do we need this ignore now?

neo-jesse replied:
PIL/Pillow is only used in tools, not in the main package:
Used in shared-data/tools/format_image.py (a tool script)
Not used in opentrons_shared_data/ (the main package code)

Pillow is only in dev dependencies:
pyproject.toml: dev = [..., 'pillow', 'types-Pillow']
Not in main dependencies or robot extra

mypy may check the tools directory:
If Pillow isn't installed, mypy would report missing imports
The ignores prevent those errors

sfoster1 reviewed (changes requested):
let's try and do all we can to avoid changes that aren't necessary for the uv cutover. that's stuff like
- Altering the `pipenv_opts` definition requiring changes to all the makefiles to push `--dev` up
- Whatever is causing the typechecking to change and requiring some extra `cast`s 
- Whatever made us reformat `ot3controller.py`

There's also some other little stuff:
- is there ever a time we would not want to install dev dependencies? can we just remove that argument to the action, or invert it to `no-install-dev-deps` and specify it more rarely?
- let's rely on python depspecs for the linux only dependencies
- some code duplication around path specification i think we can pull
- let's just double check that the versions of all the dependencies, including dev dependencies, are the same in the new uv and the old pipenv dependency specs. that will let us change fewer things at one time

y3rsh commented on .github/actions/python/install-uv/action.yaml:
I hope we can switch back to using their action.  What bug is this?

y3rsh commented on api/Makefile:
```diff
@@ -82,7 +104,7 @@ clean: docs-clean
 .PHONY: teardown
 teardown:
-	-$(pipenv) --rm
+	rm -rf .venv
```
Like we used to do with pipenv
I think here we want to use uv to resolve the location of the venv and remove it instead of resorting to rm.
`uv venv --remove` mean uv , use your resolution rules to find the venv for this project and remove it.
```suggestion
	$(UV) venv --remove
```

y3rsh commented on api/Makefile:
```diff
@@ -171,7 +193,7 @@ dev:
 .PHONY: local-shell
 local-shell:
-	$(pipenv) shell
+	uv run bash
```
```suggestion
	$(UV) run $(SHELL)
```
This might be better? 
On my mac I have `SHELL=/bin/zsh`
Prolly is set on your linux machine too.

neo-jesse committed:
squashed commits

more squashed commits

y3rsh commented on .github/actions/python/cache-uv/action.yaml:
Not necessary, 
use 
https://github.com/astral-sh/setup-uv/tree/v7/ to setup/specify python version, install uv and caching globs are used for every job, no need for custom caching.  caching globs should be contained to the pyproject.toml and uv.lock in each of the python directories listed below.

y3rsh commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,108 @@ inputs:
+    description: 'Whether to install system dependencies (libsystemd-dev on Linux)'
+    required: false
+    default: 'true'
+  install-dev-deps:
```
Is there some place that it is advantages for us to not install teh dev dependencies? I think that `make setup` by default will install the dev dependencies.  In api and shared data we will also have a robot section in the depency-groups that pins all required dependencies to the exact version that will be on the robot.  In all the other folders that do not get published to pypi the dependencies section will be pinned == dependencies that are exactly the version of the dependency on the robot.

y3rsh commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,108 @@ inputs:
+    description: 'Whether to install dev optional dependencies (UV projects only: --group dev)'
+    required: false
+    default: 'true'
+outputs:
```
go ahead and remove

y3rsh commented on .github/actions/python/setup/action.yaml:
I think you should  create a new action under actions/python 
setup-uv and not touch setup
Call that in all the workflows where you want it instead of setup
It also makes it clearer what is shifted to uv.

y3rsh commented on abr-testing/Pipfile.lock:
should not be changed

neo-jesse replied:
The abr-testing Pipfile.lock changed because abr-testing depends on opentrons and opentrons-shared-data as editable dependencies. When we downgraded jsonschema from 4.21.1 to 4.17.3 in api and shared-data

y3rsh commented on api/mypy.ini:
Should not be changed

y3rsh commented on api/Makefile:
```diff
@@ -61,19 +61,27 @@ clean_all_cmd = $(clean_cmd) dist
+	else \
+		uv sync --python 3.10 --group dev; \
+	fi
+endef
```
I think this is a remnant of getting started on this work.  There will always be a `uv.lock` and it should always be used (`--frozen`).  The group should also always be `dev` as it requires the group robot where we lock the dependencies as closely as possible to what is on the robot for all dev activity.
So this may be simplified to always be:
`uv sync --frozen --python 3.10 --group dev`

y3rsh commented on api/uv.lock.backup:
Can remove this?

y3rsh commented on hardware/pyproject.toml:
```diff
@@ -17,16 +17,36 @@ classifiers = [
 ]
 dynamic = ['version', 'hatch-dependency-coversion']
 
-[project.optional-dependencies]
-"FLEX" = ["python-can==4.2.2"]
```
This needs added back

neo-jesse replied:
I thought you didn't want "optional-dependencies"? Or just in api/shared-data?

neo-jesse replied:
Keep python-can==4.2.2 in dependencies (required because it's imported unconditionally)
Don't put it in optional-dependencies (redundant and confusing)

If you want python-can to be optional (only installed with pip install opentrons-hardware[FLEX]), you'd need to make the imports conditional (try/except or behind feature flags), but that doesn't appear to be the case here.

y3rsh commented on robot-server/Makefile:
```diff
@@ -69,16 +69,24 @@ run_dev ?= uvicorn "robot_server.app:app" --host $(dev_host) --port $(dev_port)
-	$(pipenv) run pip freeze
+	@$(uv_sync_dev)
+	@uv pip uninstall python-can
+	@uv pip list
```
I think these should be `$(UV)`

neo-jesse replied:
Updating to $(UV) everywhere, if possible

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,51 @@ inputs:
+    - name: Setup Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: ${{ inputs.python-version != 'false' && inputs.python-version || '3.10' }}
```
what does `inputs.python-version` being the literal `'false'` mean? why would we specify it?

neo-jesse replied:
good point!

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,51 @@ inputs:
+        echo "start-time=$START_TIME" >> $GITHUB_OUTPUT
+        # For pipenv projects, make setup runs "pipenv sync --dev" which always installs dev dependencies
+        # (see scripts/python.mk where pipenv_opts includes --dev)
+        make -C ${{ inputs.project }} setup
```
i think this unfortunately won't work for shared-data, which needs `setup-py`; we might want to add `setup-py` targets to the other makefiles so it can be uniform

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -8,31 +8,51 @@ inputs:
+        END_TIME=$(date +%s)
+        DURATION=$((END_TIME - START_TIME))
+        echo "duration=$DURATION" >> $GITHUB_OUTPUT
+        echo "::notice::Setup completed in ${DURATION} seconds"
```
what does the time monitoring give us that github's built in time monitoring of build steps doesn't?

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -74,7 +70,6 @@ jobs:
     strategy:
       matrix:
         os: ['windows-2022', 'ubuntu-24.04', 'macos-latest']
-        python: ['3.10']
```
let's keep the python version element of the matrix around and continue to specify it to `setup-uv`, because when we update to 3.12 I would love to keep 3.10 compatibility and we'll need this to test it.

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -141,9 +130,14 @@ jobs:
           git fetch -f origin ${{ github.ref }}:${{ github.ref }}
           git checkout ${{ github.ref }}
-      - uses: 'actions/setup-python@v4'
+      - name: Install UV
```
can we use the `setup-uv` action here? make a `setup` recipe in package-testing?

sfoster1 commented on .github/workflows/http-docs-build.yaml:
```diff
@@ -1,79 +0,0 @@
-# This workflow builds HTTP API docs when either:
```
why's this deleted? if it's something we're doing for unrelated reasons, can we do it in a different pr?

y3rsh replied:
Ya it is completely broken in edge, I told Jesse to delete to unblock.  Will fix in another PR.

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -69,10 +65,6 @@ jobs:
         os: ['windows-2022', 'ubuntu-24.04', 'macos-latest']
-        # TODO(mc, 2022-02-24): expand this matrix to 3.8 and 3.9,
-        # preferably in a nightly cronjob on edge or something
-        python: ['3.10']
```
as in the api makefile, i'd prefer to keep this matrix present so when we update to 3.12 we can try and keep 3.10 compatibility

sfoster1 commented on api/mypy.ini:
```diff
@@ -10,3 +10,20 @@ init_typed = True
 warn_required_dynamic_aliases = True
 warn_untyped_fields = True
 
+# Ignore optional/local packages that may not be installed
```
this definitely should not be changed - we have to be able to do cross-project typechecking. these are not optional and must be installed for our dev setup to work.

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,25 +19,74 @@ description = "The Opentrons API is a simple framework designed to make writing
-    "numpy>=1.20.0,<2",
-    "pydantic>=2.0.0,<3",
-    "pydantic-settings>=2,<3",
+    "aionotify~=0.3.1",
```
these are all the range dependencies for library-style installs, right? why do we need to narrow the acceptable ranges?

y3rsh replied:
Ya these just need to look exactly as they did before.
Make sure to re lock (uv lock) to update the `uv.lock`

sfoster1 commented on app-shell/Makefile:
```diff
@@ -62,22 +62,26 @@ builder := yarn electron-builder \
 	--config.electronVersion=33.2.1 \
 	--publish never
 
+# Resolve the API python interpreter lazily so `make setup` (and other targets
```
this is a great change, thank you!

sfoster1 commented on robot-server/pyproject.toml:
```diff
@@ -30,46 +30,109 @@ dependencies = [
+    "mypy.ini",
+    "pytest.ini",
+    ".flake8",
+    # Pipenv artifacts intentionally omitted; we're on uv now
```
better as a commit message than a checked-in comment

sfoster1 commented on scripts/python.mk:
```diff
@@ -1,22 +1,12 @@
-# This environment variable can be used to select a specific
```
if we're not using this file anymore, can we just delete it?

neo-jesse replied:
We still use it in a few places - python_package_version, python_get_wheelname, hardware-testing, abr-testing, and test-data-generation.

sfoster1 commented on shared-data/pyproject.toml:
```diff
@@ -20,12 +20,45 @@ classifiers = [
-    "jsonschema>=4.0.0,<5",
     "typing-extensions>=4.0.0,<5",
-    "pydantic>=2.0.0,<3.0.0",
+    "pydantic~=2.11.7",
```
why do we need to narrow the version range here?

neo-jesse replied:
this change matches it to api/robot-server, which also depend on pydantic>=2.11.7. When we migrated everything to uv, older pins like ~=2.6 or 2.9 caused resolver conflicts—opentrons (API) requires >=2.11.7,<2.12, so any transitive consumer had to match it. Updating shared-data eliminates that clash: uv can resolve a single pydantic version for all packages, and CI/tests no longer fail when opentrons and shared-data get installed together.

sfoster1 commented on shared-data/pyproject.toml:
```diff
@@ -20,12 +20,45 @@ classifiers = [
 dynamic = ['version']
 
+[dependency-groups]
+# Robot dependencies - exact versions from Pipfile [packages] for robot deployment
```
that it's from the pipfile won't be relevant over the long term, better as a commit message than a checked-in comment. but the "robot dependencies for robot deployment" is a great thing to keep

sfoster1 commented on shared-data/pyproject.toml:
```diff
@@ -20,12 +20,45 @@ classifiers = [
+    "numpy==1.22.3",
+    "zipp==3.21.0",
+]
+# Dev dependencies - from Pipfile [dev-packages]
```
see above re: the "from pipfile" part

sfoster1 reviewed (changes requested):
Very much getting there! A couple little nits to pick; the only big things left are some github actions stuff, the thing with the mypy configuration in `api`, and a weirdness with the dependencies in `api`.

sfoster1 commented on .github/actions/python/setup-uv/action.yaml:
```diff
@@ -0,0 +1,40 @@
+  python-version:
+    description: 'Python version to install via setup-uv'
+    required: false
+    default: '3.10'
```
```suggestion
    default: '3.12'
```

sfoster1 commented on .github/actions/python/setup/action.yaml:
```diff
@@ -4,35 +4,39 @@ inputs:
+    - name: Setup Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: '3.10'
```
```suggestion
        python-version: '3.12'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -55,16 +55,13 @@ jobs:
+      - uses: './.github/actions/python/setup-uv'
         with:
           project: 'api'
+          python-version: '3.10'
```
```suggestion
          python-version: '3.12'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -55,16 +55,13 @@ jobs:
         with:
           fetch-depth: 0
-      - uses: 'actions/setup-node@v6'
+      - uses: 'actions/setup-node@v4'
```
```suggestion
      - uses: 'actions/setup-node@v6'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -74,7 +71,7 @@ jobs:
       matrix:
         os: ['windows-2022', 'ubuntu-24.04', 'macos-latest']
-        python: ['3.12']
+        python: ['3.10']
```
```suggestion
        python: ['3.12']
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -92,27 +89,22 @@ jobs:
           git fetch -f origin ${{ github.ref }}:${{ github.ref }}
           git checkout ${{ github.ref }}
-      - uses: 'actions/setup-node@v6'
+      - uses: 'actions/setup-node@v4'
```
```suggestion
      - uses: 'actions/setup-node@v6'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -129,16 +121,10 @@ jobs:
     strategy:
       matrix:
         os: ['ubuntu-24.04', 'macos-latest', 'windows-2022']
-        python: ['3.10', '3.11', '3.12', '3.13']
```
undo this change, keep the matrix and the exclude

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -147,21 +133,12 @@ jobs:
+        uses: './.github/actions/python/setup-uv'
+        with:
+          project: 'package-testing'
+          python-version: '3.10'
```
```suggestion
          python-version: '3.12'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -205,25 +182,22 @@ jobs:
           git fetch -f origin ${{ github.ref }}:${{ github.ref }}
           git checkout ${{ github.ref }}
-      - uses: 'actions/setup-node@v6'
+      - uses: 'actions/setup-node@v4'
```
```suggestion
      - uses: 'actions/setup-node@v6'
```

sfoster1 commented on .github/workflows/api-test-lint-deploy.yaml:
```diff
@@ -205,25 +182,22 @@ jobs:
+      - uses: './.github/actions/python/setup-uv'
         with:
           project: 'api'
+          python-version: '3.10'
```
```suggestion
          python-version: '3.12'
```

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -54,13 +54,10 @@ jobs:
+      - uses: './.github/actions/python/setup-uv'
         with:
           project: 'shared-data'
+          python-version: '3.10'
```
```suggestion
          python-version: '3.12'
```

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -73,6 +70,7 @@ jobs:
         os: ['windows-2022', 'ubuntu-24.04', 'macos-latest']
         python: ['3.12']
 
+        python: ['3.10']
```
remove

sfoster1 commented on .github/workflows/shared-data-test-lint-deploy.yaml:
```diff
@@ -172,12 +167,10 @@ jobs:
+      - uses: './.github/actions/python/setup-uv'
         with:
           project: 'shared-data'
+          python-version: '3.10'
```
```suggestion
          python-version: '3.12'
```

sfoster1 commented on package-testing/setup.ps1:
```diff
@@ -10,20 +10,22 @@ if (Test-Path -Path $VENV_DIR) {
 
 Write-Output "Creating virtual environment in $VENV_DIR..."
-python -m venv $VENV_DIR
+uv venv $VENV_DIR --python 3.10
```
```suggestion
uv venv $VENV_DIR --python 3.12
```

sfoster1 commented on package-testing/setup.sh:
```diff
@@ -11,25 +11,27 @@ if [ -d "$VENV_DIR" ]; then
 
 echo "Creating virtual environment in $VENV_DIR..."
-python -m venv "$VENV_DIR"
+uv venv "$VENV_DIR" --python 3.10
```
```suggestion
uv venv "$VENV_DIR" --python 3.12
```

neo-jesse committed:
suggestions

neo-jesse committed:
numpy update

neo-jesse committed:
python-uv got removed

neo-jesse committed:
wrong version

neo-jesse committed:
more numpy updates

neo-jesse committed:
uv lock

neo-jesse committed:
uv lock, numpy version again

neo-jesse committed:
more numpy

neo-jesse committed:
no 3.10

neo-jesse committed:
regen lock files with 3.12

ecormany committed:
docs: update macOS and Ubuntu required versions (#20355)

# Overview

Update outdated operating system requirements in the Flex instruction
manual.

Closes RTC-666 😈 

## Test Plan and Hands on Testing

Sandbox for
[each](https://sandbox.docs.opentrons.com/docs-update-os-reqs/flex/opentrons-app/#app-installation)
[page](https://sandbox.docs.opentrons.com/docs-update-os-reqs/flex/system-description/specs/#general-specifications).

## Changelog

- macOS 10.10 -> 10.16
- Ubuntu 12.04 -> 20.04

## Review requests

Search across all mkdocs source doesn't indicate any other occurrences,
but let's be sure.

## Risk assessment

nil

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
 
 dependencies = [
    "opentrons-shared-data==0.0.0",
+    "aionotify==0.3.1",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
 dependencies = [
    "opentrons-shared-data==0.0.0",
+    "aionotify==0.3.1",
+    "anyio>=4.9.0,<5.0.0",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
    "opentrons-shared-data==0.0.0",
+    "aionotify==0.3.1",
+    "anyio>=4.9.0,<5.0.0",
+    "jsonschema>=3.0.1,<4.18.0",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
+    "aionotify==0.3.1",
+    "anyio>=4.9.0,<5.0.0",
+    "jsonschema>=3.0.1,<4.18.0",
+    "numpy>=1.26.3,<2",
```
```suggestion
    "numpy>=1.26.4,<2",
```
ended up being 4

sfoster1 commented on api/pyproject.toml:
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
+    "anyio>=4.9.0,<5.0.0",
+    "jsonschema>=3.0.1,<4.18.0",
+    "numpy>=1.26.3,<2",
+    "pydantic>=2.0.0,<3",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -19,6 +19,12 @@ description = "The Opentrons API is a simple framework designed to make writing
+    "jsonschema>=3.0.1,<4.18.0",
+    "numpy>=1.26.3,<2",
+    "pydantic>=2.0.0,<3",
+    "pydantic-settings>=2,<3",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -29,15 +35,65 @@ dependencies = [
     "typing-extensions>=4.0.0,<5",
     "click>=8.0.0,<9",
     "pyusb==1.2.1",
+    "pyusb==1.2.1",
```
```suggestion
```
duplicate

sfoster1 commented on api/pyproject.toml:
```diff
@@ -29,15 +35,65 @@ dependencies = [
+    "pyusb==1.2.1",
     'importlib-metadata >= 1.0 ; python_version < "3.8"',
-    "packaging>=21.0",
+    "packaging>=22.0",
```
```suggestion
    "packaging>=25.0",
```

sfoster1 commented on api/pyproject.toml:
```diff
@@ -29,15 +35,65 @@ dependencies = [
+    "pydantic-settings==2.4.0",
+    "anyio==4.9.0",
+    "numpy==1.26.4",
+    "packaging==22.0",
```
```suggestion
    "packaging==25.0",
```

sfoster1 commented on robot-server/pyproject.toml:
```diff
@@ -30,47 +30,109 @@ dependencies = [
 ]
 
-dynamic = ['version', 'hatch-dependency-coversion']
+[dependency-groups]
```
we need a robot dependency group here i believe

sfoster1 commented on system-server/pyproject.toml:
```diff
@@ -16,16 +16,52 @@ classifiers = [
 
 dynamic = ['version']
 
+[dependency-groups]
```
need a robot dependency group here

sfoster1 commented on update-server/pyproject.toml:
```diff
@@ -19,14 +19,34 @@ classifiers = [
 ]
 dynamic = ['version']
 
+[dependency-groups]
```
need a dependency group here

sfoster1 commented on usb-bridge/pyproject.toml:
```diff
@@ -22,10 +23,34 @@ classifiers = [
 dynamic = ["version"]
-readme = "README.md"
+
+[dependency-groups]
```
need a robot dependency group here

sfoster1 commented on scripts/python-uv.mk:
```diff
@@ -0,0 +1,99 @@
+	$(UV) sync --frozen --group dev --python $(UV_PYTHON)
+endef
+
+poetry := poetry
```
do we need these?

sfoster1 commented on hardware/pyproject.toml:
```diff
@@ -17,17 +17,41 @@ classifiers = [
 ]
 keywords = ["robots", "protocols", "synbio", "pcr", "automation", "lab"]
 description = "Hardware control for Opentrons Robots."
+readme = "README.md"
```
this needs a lockfile

neo-jesse committed:
should be using UV already

koji committed:
feat(app): add close action to second window (#20351)

* feat(app): add close action to second window

neo-jesse committed:
typeguard update

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml remove pydantic

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
removed poetry defs

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update api/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
robot dependencies

y3rsh commented on robot-server/pyproject.toml:
```diff
@@ -30,47 +30,127 @@ dependencies = [
+    "sqlalchemy==1.4.51",
+    "paho-mqtt==1.6.1",
+    "aiohttp==3.12.14",
+]
```
I do not understand why we would need to repeat 
`project.dependencies`
in  
`project.dependency-groups.robot`
in a folder like robot-server that is not a published package, `project.dependencies` should always be the robot dependencies specified with == and matching the version of the dependency across the repo if shared.
### Only
in /api and /shared-data do we `project.dependency-groups.robot` because the `project.dependencies` are ranges for the package, but everywhere else, `project.dependencies` are the robot dependencies.

sfoster1 replied:
Oh, sure, we can do that too if we want. just put the exact deps in the regular dependencies. imo if we do that we should just have a copy of the defaults named `robot` for consistency and ease of openembedded integration

neo-jesse replied:
OK, it is in both places now for these 4 projects.

rclarke0 committed:
refactor(app): use OffsetTag in HistoricalProtocolRunDrawer (#20356)

# Overview

Uses OffsetTag to wrap LPC offsets in `HistoricalProtocolRunDrawer`

## Test Plan and Hands on Testing

smoke tested on the app
<img width="873" height="646" alt="Screenshot 2025-12-11 at 3 11 44 PM"
src="https://github.com/user-attachments/assets/c1b25d09-5eb6-4dc3-9511-86df886099bf"
/>

## Changelog

- replaced `LegacyOffsetVector` with `OffsetTag`

## Risk assessment

low

Closes EXEC-1891

sfoster1 commented:
Okay one last thing is whenever we change the dependencies at all, even to add a duplicate group, we have to relock and check in `uv.lock` - otherwise tools that use only the lockfile, like OE's use of `uv export ... --frozen` won't see the changes

neo-jesse committed:
uv.lock updates

ryanthecoder committed:
feat(hardware-testing): Create a protocol version of belt calibration (#20281)

# Overview
This protocol does all the same stuff as the previous hardware testing
script except it does not have any dependencies on hardware testing. It
also does not export a CSV report since this is only used in the factory
and this is meant for use in the field during preventative maintenance.

ddcc4 committed:
refactor(protocol-designer): help track down why _formLevelErrors() is crashing (#20361)

# Overview

I can't figure out why we're seeing these crashes in
`_formLevelErrors()`:
- https://opentrons-76.sentry.io/issues/7079511965?project=4509363266387968
- https://opentrons-76.sentry.io/issues/7079511914?project=4509363266387968
- https://opentrons-76.sentry.io/issues/7079512035?project=4509363266387968

The failure can happen if our `getFormErrors()` doesn't handle a form
`stepType`, but as far as I can tell, all possible `stepType`s are
already handled.

It could also happen if the saved protocol is malformed (I was able to
trigger the same exception if I manually edited a protocol to say
`"stepType": "moveLiquidxxx"`), or maybe people are importing newer
protocols with new `stepType`s into older versions versions of PD? It's
also possible we're creating a `HydratedFormData` somewhere without
setting `stepType`, which would also trigger this error if `stepType` is
`undefined`.

This PR just adds some instrumentation so that we get a more useful
error message when this crash happens again.

## Test Plan and Hands on Testing

Ran PD locally.

## Risk assessment

Low.

ecormany committed:
chore: corrections to 8.8.0 release notes (#20362)

sfoster1 committed:
fix(api): handle new flex ethernet iface (#20354)

It's called end0 now for some inscrutable systemd reason.

rclarke0 committed:
feat(abr-testing): track image captures and attach image to error tickets (#20325)

# Overview

Track image captures in abr data sheet
Attached images from protocol to error tickets and completed run slack
message
Attaches last 3 seconds of the livestream to error run slack message

## Test Plan and Hands on Testing

**smoke tested on abr-run-data tracker**
<img width="1069" height="163" alt="Screenshot 2025-12-08 at 3 22 52 PM"
src="https://github.com/user-attachments/assets/c6722997-9677-4774-bd3c-ac71bfb84a07"
/>
**smoke tested with jira error script:**
<img width="935" height="551" alt="Screenshot 2025-12-08 at 3 24 01 PM"
src="https://github.com/user-attachments/assets/484039b9-0a4c-44fe-b949-2073bafcf78e"
/>
**smoke tested slack messages**
<img width="502" height="574" alt="Screenshot 2025-12-11 at 1 12 47 PM"
src="https://github.com/user-attachments/assets/6486c1b1-6341-49a9-8773-9791a77585a8"
/>


## Risk assessment

low

koji committed:
fix(app): fix small bugs in Second Window (#20360)

* fix(app): fix small bugs in Second Window

SyntaxColoring committed:
refactor(step-generation): Split up tests into local subdirectories (#20365)

## Overview

step-generation's source is organized into a bunch of subdirectories,
like `/utils/`, `/commandCreators/atomic/`,
`/commandCreators/compound/`, etc., but its unit tests were all living
together in one flat top-level directory. With a bunch of
similarly-named test files, that was getting difficult to follow.

As discussed in
https://github.com/Opentrons/opentrons/pull/20349#discussion_r2610861235,
this splits the monolithic `__tests__` directory into a bunch of
distributed "local" `__test__` directories, following the pattern
elsewhere in our codebase.

## Test Plan and Hands on Testing

* [x] Compared the test count before and after to make sure I didn't
accidentally remove anything.

koji committed:
fix(app) add ModuleSlot for Second Window (#20366)

* fix(app) add ModuleSlot for Second Window

y3rsh commented on scripts/python-uv.mk:
```diff
@@ -0,0 +1,91 @@
+# UV is required for Python dependency management
+UV ?= uv
+UV_PYTHON = 3.12
+
```
On my local system where I have many pythons managed by uv, when I tried, 
`make -C robot-server dev` I would have issues because even with .python-version file at 3.12 at the root, 3.12 was not being used to spin up the robot server.

I had to solve that by setting it on the make invocation
`make UV_PYTHON=3.12 -C robot-server dev` 

If we export here, recipe commands (including ones that invoke uv without --python $(UV_PYTHON)) like `make -C robot-server dev` will work without having to set it each time.

```suggestion
export UV_PYTHON
```

sfoster1 replied:
sure, that sounds good. I think we had something similar in the old system with `OT_PYTHON` too, we could keep that around

sfoster1 commented:
Successful flex build: https://github.com/Opentrons/oe-core/actions/runs/20175271922

neo-jesse committed:
Add UV_PYTHON export

Co-authored-by: Josh McVey <josh.mcvey@opentrons.com>

sfoster1 commented on api/pyproject.toml:
```diff
@@ -30,14 +31,63 @@ dependencies = [
 
 dynamic = ['version', 'hatch-dependency-coversion']
 
-[project.optional-dependencies]
```
when you do a `uv lock` of the robot server, uv will see that the robot server needs `opentrons` and look here for it; and it will scan this project for transitive dependencies... but _only_ the transitive dependencies that will show up when this is packaged, e.g. `project.dependencies`. So right now, the robot server lockfile doesn't have `opentrons_hardware` in it at all.

There's a couple things we could do with this but I think the easiest one is to add these deps to the robot server's `robot` dependency group.

sfoster1 commented on api/pyproject.toml:
```diff
@@ -30,14 +31,63 @@ dependencies = [
+    "packaging==25.0",
+    "pyusb==1.2.1",
+    "zipp==3.21.0",
+    "opentrons-hardware==0.0.0",
```
```suggestion
    "opentrons-hardware[FLEX]==0.0.0",
```

sfoster1 commented on robot-server/pyproject.toml:
```diff
@@ -30,47 +30,127 @@ dependencies = [
+[dependency-groups]
+# Robot dependencies - exact versions for robot deployment
+robot = [
+    "opentrons==0.0.0",
```
```suggestion
    "opentrons==0.0.0",
    "opentrons-hardware[FLEX]==0.0.0",
```

sfoster1 commented on robot-server/pyproject.toml:
```diff
@@ -30,47 +30,127 @@ dependencies = [
     "uvicorn==0.27.0.post1",
     "wsproto==1.2.0",
-    "systemd-python==234; sys_platform=='linux'",
+    "systemd-python==234; sys_platform == 'linux'",
```
```suggestion
    "opentrons-hardware[FLEX]==0.0.0",
    "systemd-python==234; sys_platform == 'linux'",
```

sfoster1 reviewed (changes requested):
Need to get the `opentrons_hardware` dep in here

neo-jesse committed:
Adding FLEX

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update robot-server/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
Update robot-server/pyproject.toml

Co-authored-by: Seth Foster <seth@opentrons.com>

neo-jesse committed:
uv lock changes

koji commented on DEV_SETUP.md:
````diff
@@ -148,7 +150,29 @@ eval "$(pyenv init -)"
+echo 'export PATH="$HOME/.local/bin:$PATH"' >> ~/.zprofile
+```
+
+#### 4. Install `jpeg` if on ARM Mac (M1/M2/M3)
````
```suggestion
#### 4. Install `jpeg` if on ARM Mac (M1/M2/M3/M4)
```

koji committed:
chore: update prettier to v3.7.0 (#20373)

* chore: update prettier to v3.7.0

dependabot[bot] committed:
chore(deps): bump @thi.ng/paths from 1.6.5 to 5.1.63 in /app (#20372)

* chore(deps): bump @thi.ng/paths from 1.6.5 to 5.1.63 in /app

koji committed:
fix(components): fix prop name (#20371)

* fix(components): fix prop name

SyntaxColoring committed:
fix(protocol-designer): Align number to top of step button instead of middle (#20370)

neo-jesse committed:
Comment update

SyntaxColoring committed:
refactor(protocol-designer): Deduplicate module state types with step-generation (#20369)

step-generation and protocol-designer define identical types and
constants for module state. They will get more complicated with
concurrent module actions, so we gotta deduplicate 'em.

Goes towards EXEC-2009.

rclarke0 committed:
refactor(app): update logic to determine which protocol set up step is next (#20364)

# Overview

Determines the text of the next step button using a switch statement
rather than a ternary operator to include all possible next steps

## Test Plan and Hands on Testing

- Tested protocols where the next step would be modules, LPC, labware,
or camera
attached videos to ticket EXEC-1434

<img width="914" height="236" alt="Screenshot 2025-12-12 at 1 00 03 PM"
src="https://github.com/user-attachments/assets/72ca71f5-4bcd-44f3-824e-5592981d3a89"
/>
<img width="892" height="248" alt="Screenshot 2025-12-12 at 1 00 14 PM"
src="https://github.com/user-attachments/assets/b98335fb-cfa7-4e72-9820-0c5d79e8575d"
/>
<img width="900" height="318" alt="Screenshot 2025-12-12 at 1 00 26 PM"
src="https://github.com/user-attachments/assets/b8670124-ab4a-4460-babe-b8394b4c13ad"
/>
<img width="913" height="248" alt="Screenshot 2025-12-12 at 1 00 34 PM"
src="https://github.com/user-attachments/assets/a102b06e-7838-4b2b-9daf-54ffa5e07daf"
/>

## Changelog

- checks if labware or liquids are loaded in protocolAnalysis to
determine if `Labware & Liquids` confirmation is needed
- creates new type `AnalyticsProtocolProceedButtonText` to define what
the next step button text should be
- adds modules setup step to the list of applicable steps if fixtures or
modules are not ready in `ProtocolRunSetup`. The hook if module set up
is included is based on if their are modules are in the protocol. This
does not always address if the deck hardware should be adjusted.
- 
## Risk assessment

- medium - determines button text and which protocol setup steps are
needed

Closes EXEC-1434

koji committed:
fix(app, protocol-designer): remove background color from PrimaryButton (#20377)

* fix(app, protocol-designer): remove background color from PrimaryButton

koji committed:
fix: remove alert-circle (#20378)

* fix: remove alert-circle

koji committed:
chore(app-shell, app-shell-odd): remove thi.ng/paths (#20382)

* chore(app-shell, app-shell-odd): remove thi.ng/paths

koji committed:
chore(opentrons-ai-client): remove unused component (#20383)

* chore(opentrons-ai-client): remove unused component

ncdiehl11 committed:
feat(protocol-designer,-step-generation): build stacker step UI (#20367)

This PR fleshes out the flex stacker step form content. It also
refactors some of our Redux infrastructure during container creations in
deck setup to update the flex stacker's state.

mjhuff committed:
refactor(app): update zipfile naming convention (#20380)

Closes EXEC-2134

mjhuff committed:
refactor(app): update camera status copy conditionally (#20376)

Closes EXEC-2132

y3rsh committed:
fix(pd): source map creation and docs (#20381)

koji committed:
fix(app): modify error step layout (#20374)

* fix(app): modify error step layout

dependabot[bot] committed:
chore(deps-dev): bump vitest from 2.1.8 to 2.1.9 (#20386)

* chore(deps-dev): bump vitest from 2.1.8 to 2.1.9 (#20386)

jbleon95 committed:
feat(shared-data): command annotations schema v2 (#20324)

Adds a new command annotation schema to shared data, adds this type to the existing protocol schema v8, and makes a small change in the previous and mostly unused version of command annotations to not break if it somehow encounters the new command annotations.

skowalski08 committed:
test(pd-e2e): PD E2E Move Labware and Lids (#20363)

# Overview

Added playwright testing for moving Lids (TC and Universal) around the
deck manually or with the gripper.
Addresses Phase 1: RQA-4958 & Phase 2: RQA-4951

## Test Plan and Hands on Testing

E2E tests

## Changelog

Added tests to check Move Labware on Flex Protocol
Added functions that allow you to easily add move labware with or
without gripper

## Review requests

- Are selectors strong enough? (I did use codegen)
- There are two notes on test_pd_move_labware, search "###NOTE" and you
will see my comments

## Risk assessment

- Selectors likely fragile

sfoster1 commented:
this build boots up ok! https://github.com/Opentrons/oe-core/actions/runs/20271933785

rclarke0 committed:
refactor(protocol-designer, step-generation): remove remaining maxPoolCount prop (#20387)

# Overview

Removes remaining mentions of `maxPoolCount` since we can get this
property from the hopper's set stored labware

## Test Plan and Hands on Testing

ran unit tests and all passed!

## Risk assessment

low

neo-jesse committed:
fix(ci, dependabot): forks can't access secrets (#20389)

### Skip deployment for fork PRs (Dependabot)

Problem:
Dependabot PRs (from forks) fail on deploy jobs because they can't
access secrets/OIDC tokens for AWS credentials.

Solution:
Added conditions to skip deployment jobs for fork PRs while still
running tests and builds.

Changes:
Updated deploy job conditions in 5 workflows to check if PR is from a
fork:
components-test-build-deploy.yaml
ll-test-build-deploy.yaml
pd-test-build-deploy.yaml
docs-build.yaml
docs-build-deploy.yaml

Result:
Tests and builds run on all PRs (including Dependabot)
Deployment only runs for same-repo PRs and push events
Fork PRs (Dependabot) skip deployment instead of failing

Condition added:
`(github.event_name != 'pull_request' ||
github.event.pull_request.head.repo.full_name == github.repository)`

rclarke0 committed:
feat(api-client, app, react-api-client): bindings and implementation for previewCaptureImage (#20379)

# Overview

Create bindings for `previewCaptureImage` and implement in the app

## Test Plan and Hands on Testing

- console logged that post request was being sent with the correct
settings:
<img width="1497" height="589" alt="Screenshot 2025-12-15 at 12 33
18 PM"
src="https://github.com/user-attachments/assets/849db6da-9b84-4fba-a280-371c90278f3f"
/>

- updated unit tests

## Review requests


## Risk assessment

medium

koji committed:
docs: update python version in DEV_SETUP (#20394)

* docs: update python version in DEV_SETUP

sfoster1 committed:
chore(system-server): fix hanging tests (#20395)

Tests were hanging (more often in ci, but also locally) because
sometimes the dev server for the tavern backend would fail to start
because a port was in use. Use a random port instead.

Probably better would be letting uvicorn pick which one to use, but it's
a bit of a pain to figure out which one it picked so don't.

jerader committed:
refactor(protocol-designer): various deck setup flex stacker feature (#20357)

closes EXEC-2121

---------

Co-authored-by: ncdiehl11 <ndiehl@opentrons.com>

y3rsh committed:
chore(static-deploy): cleanup and document new process (#19815)

SyntaxColoring committed:
refactor(protocol-designer): Delete unused properties in Thermocycler substeps (#20391)

## Overview

In our UI for showing substeps, Thermocycler profiles are peculiar in that we don't actually get data from `ThermocyclerProfileSubstepItem`. Instead, we get it from the step form. This means the properties of `ThermocyclerProfileSubstepItem` are unused by anything.

The properties of `ThermocyclerProfileSubstepItem` would need to update in #20390. Since they're unused, it seems easier to just delete them instead of updating them.

This goes towards EXEC-2009.

koji committed:
refactor(components,protocol-designer): refactor Flex component (#19266)

* refactor(components,protocol-designer): refactor Flex component

koji committed:
fix(step-generation): fix white screen from def undefined (#20384)

* fix(step-generation): fix white screen from def undefined

koji committed:
fix(app): fix CSS Modules migration bugs buttons (#20403)

* fix(app): fix CSS Modules migration bugs buttons

sfoster1 reviewed (changes requested):
let's add `setuptools` as a dev dependency a la https://github.com/Opentrons/opentrons/pull/20402#pullrequestreview-3587996415

rclarke0 committed:
feat(protocol-designer): mixpanel analytics for stacker (#20396)

# Overview

Tracks flex stacker step creation, labware loaded into the stacker with
liquids, and how many stackers are loaded per protocol.

## Test Plan and Hands on Testing

- Added log statements to liquids step to test
- Will need to smoke test stacker commands once step-generation is
complete.

Liquids log:
<img width="803" height="256" alt="Screenshot 2025-12-16 at 1 56 43 PM"
src="https://github.com/user-attachments/assets/3a2df9a5-fffb-4a14-a142-dc5b466239f1"
/>

## Risk assessment

low

closes EXEC-1546

koji committed:
fix(components): fix Link component text wrap (#20404)

* fix(components): fix Link component text wrap

ddcc4 committed:
refactor(protocol-designer): figure out why SecondStepsMoveLiquidTools is crashing (#20401)

# Overview

`<SecondStepsMoveLiquidTools>` is failing somehow because it can't find
the requested liquid class:

https://opentrons-76.sentry.io/issues/6987511353?project=4509363266387968

The liquid class we want comes from `formData.liquidClass`, so:

- There could be an invalid liquid class name in the form somehow,
- Or it's undefined,
- Or the `NONE_LIQUID_CLASS_NAME` -> `WATER_LIQUID_CLASS_NAME` fallback
is broken?

This PR tries to catch the error with a more informative error message
to help us figure out what's going on.

## Test Plan and Hands on Testing

Smoke-tested locally with a Transfer step.
Run CI tests.

## Risk assessment

Low.

rclarke0 committed:
fix(app): copy changes to protocol set up button (#20393)

# Overview

Copy fixes to protocol setup buttons.
finishes EXEC-1434

## Test Plan and Hands on Testing

smoke tested:
<img width="910" height="272" alt="Screenshot 2025-12-16 at 12 12 43 PM"
src="https://github.com/user-attachments/assets/72edfc19-5e5f-4c86-99a3-ed67f96d56b0"
/>
<img width="890" height="581" alt="Screenshot 2025-12-16 at 12 13 11 PM"
src="https://github.com/user-attachments/assets/440aa2de-1973-4253-8972-b7560c77a10e"
/>


## Changelog

- removed navigate to liquids text


## Risk assessment

low

koji committed:
chore(app-shell, app-shell-odd) update electron version to v39 (#20117)

* chore(app-shell, app-shell-odd) update electron version to v39

neo-jesse committed:
setuptools fix

sfoster1 reviewed (approved):
I think this is done, and I'm happy to merge it! Thank you so much for working on this and getting us off pipenv at long last.

Looks like there are some remaining conflicts; these are pipenv changes from adding setuptools (and we can just delete those files) and coincidental conflicts from updating the node version in the github actions, just take both sets of changes.

rclarke0 committed:
refactor(shared-data): handle multiple liquids in one well by combining them into a third mixed liquid (#20359)

# Overview

When multiple liquids are loaded into the same well, our current
behavior is to layer them on top of each other. This creates an unclear
UI because the shared well is always highlighted.

This PR creates a third liquid by combining the liquids in one well into
a command seperated list. The total volume is displayed.

## Test Plan and Hands on Testing
- wrote unit tests.
- smoke tested with protocol in ticket RQA-4799

## Changelog


https://github.com/user-attachments/assets/da157f7f-6f60-49cd-be24-d34dfafef780
Smoke tested with two mixed liquids, one being a mix of 3 and one being
a mix of 2
<img width="734" height="490" alt="Screenshot 2025-12-17 at 12 35 33 PM"
src="https://github.com/user-attachments/assets/8c5a31a0-a630-4329-b446-2ccb00733168"
/>


## Review requests

- I think combining liquid volume into total volume creates a confusing
experience for the user

## Risk assessment

- high changes the logic to parse through liquid loads to display in
deck setup
- this creates a different behavior than in PD - which replaces the
previous liquid if you load another one into the same well

koji committed:
chore update node version (#20408)

* chore update node version

jerader committed:
refactor(step-generation, protocol-designer): refactor empty & store state fns (#20388)

# Overview

Needed to update the labware state for the empty and store flex stacker
state functions 😄 small fixes to both

also introduces the `SYSTEM_LOCATION` labware location when you empty
something from the hopper and filters out that location in the dropdown
options for moveLiquid and moveLabware

## Test Plan and Hands on Testing

test that you can empty a labware on the flex stacker and the labware
disappears on the hopper
test that you can also store a labware and it moves to the hopper
location > NOTE this one does super work right now since you can't add a
labware to the shuttle location properly? Should be fixed by Nick's work
though that he is currently working on

## Changelog

some fixes with updating robot state in the 2 state updates


## Risk assessment

low

ncdiehl11 committed:
feat(protocol-designer,step-generation): create and wire up `HopperLabwareRenders` in `DeckSetupDetails` (#20410)

This PR adds a special-cased labware render for the top-most labware
group of a hopper, along with its potential highlight and labware
controls components.

known issues for followup:
- shuttle state is not properly updated with moveLabware step

koji committed:
fix(components, app): fix pipette container chip and add backgroundColor for web (#20405)

* fix(components, app): fix pipette container chip and add backgroundColor for we

SyntaxColoring committed:
refactor(step-generation): Update the Thermocycler state shape to allow keeping track of ongoing profiles (#20358)

## Overview

This is another step towards EXEC-2009.

On the road to getting step-generation to support concurrent
Thermocycler profiles, this lets it represent the fact that a profile is
currently in progress (i.e., a profile has been started and has not yet
been awaited).

## Changelog

`ThermocyclerModuleState` is expanded:

* The `blockTargetTemp` property is generalized. It's now
`currentBlockActivity`.
  
  Formerly, `blockTargetTemp` could only represent "the block is targeting
a temperature" and "the block is deactivated." It can now additionally
represent "the block is running a Thermocycler profile."
  
  Nothing sets the state to "the block is running a Thermocycler profile"
yet, nor does anything look for it. We're just updating the type to
allow it.

* Not only do we store the fact that a profile is running, but also the
steps of the profile, and the Protocol Engine task ID that it was
started with. In subsequent PRs, this information will help us when we
process `waitForTask` commands and need to update the Thermocycler state
to end the profile.

* A new property, `numProfilesStarted`, counts the number of times a
profile has been started on the Thermocycler. This is going to be used
to generate unique Python variable names, like `task_1 =
thermocycler.start_execute_profile(...)`, `task_2 =
thermocycler_start_execute_profile(...)`, etc.

rclarke0 committed:
feat(protocol-designer): View labware in hopper at end of run (#20409)

# Overview

Enables view labware hover text for labware in the hopper at the end of
the run

Also updated css style to ensure you can scroll through the list of
labware. Prior to this change, the labware list got cut off.

## Test Plan and Hands on Testing

- smoke tested labware viewing at start of run and end of run


https://github.com/user-attachments/assets/6dcdd48f-da70-44b4-8d4e-43096e60015c


## Risk assessment

low

Closes EXEC-2140

---------

Co-authored-by: Jethary Alcid <66035149+jerader@users.noreply.github.com>

ncdiehl11 committed:
feat(protocol-designer,-step-generation): update stacker shuttle state for move labware (#20412)

This PR wires up stacker shuttle state updates for labware stack moves
to and from the flex stacker shuttle.

Closes EXEC-2145

jerader committed:
fix(protocol-designer): slot details & labware card quantity fixes (#20413)

koji committed:
fix(app): fix the footer height for RobotUpdateProgressModal (#20414)

* fix(app): fix the footer height for RobotUpdateProgressModal

neo-jesse committed:
Merge branch 'edge' into api-and-shared-uv-migration

sfoster1 reviewed (approved):
Let's do it!!!!
